File: app/src/main/java/io/github/pshegger/infobook/InfoBookApplication.java
package io.github.pshegger.infobook;

import android.app.Application;
import android.content.SharedPreferences;
import android.preference.PreferenceManager;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.squareup.okhttp.Interceptor;
import com.squareup.okhttp.OkHttpClient;
import com.squareup.okhttp.Request;
import com.squareup.okhttp.Response;

import java.io.IOException;
import java.util.Random;

import retrofit.GsonConverterFactory;
import retrofit.Retrofit;

/**
 * Created by pshegger on 2015.10.04..
 */
public class InfoBookApplication extends Application {
    public static Random RNG;

    private static InfoBookApplication instance;
    private GWApiService mApiService;

    private String mAPIKey;

    public static InfoBookApplication getInstance() {
        return instance;
    }

    @Override
    public void onCreate() {
        super.onCreate();

        RNG = new Random();

        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(this);

        mAPIKey = prefs.getString(
                Constants.Preferences.API_KEY,
                "3674D0F3-B7A8-2242-B4C6-FE2E065C86A84C0DB899-DD85-490C-9986-1DA8E629CD21"
        );

        instance = this;
        initRetrofit();
    }

    private void initRetrofit() {
        OkHttpClient client = new OkHttpClient();
        client.interceptors().add(new Interceptor() {
            @Override
            public Response intercept(Chain chain) throws IOException {
                Request.Builder builder = chain.request().newBuilder();

                builder.addHeader("Authorization", "Bearer " + mAPIKey);

                return chain.proceed(builder.build());
            }
        });

        Gson gson = new GsonBuilder()
                .setDateFormat("yyyy-MM-dd'T'HH:mm:ssZ")
                .create();

        Retrofit retrofit = new Retrofit.Builder()
                .client(client)
                .baseUrl(Constants.GW_API_URL)
                .addConverterFactory(GsonConverterFactory.create())
                .build();

        mApiService = retrofit.create(GWApiService.class);
    }

    public GWApiService getService() {
        return mApiService;
    }

    public String getAPIKey() {
        return mAPIKey;
    }

    public void setAPIKey(String key) {
        mAPIKey = key;
        PreferenceManager.getDefaultSharedPreferences(this)
                .edit()
                .putString(
                        Constants.Preferences.API_KEY,
                        key
                ).apply();

        initRetrofit();
    }
}

File: app/src/main/kotlin/io/github/pshegger/infobook/InfoBookApplication.kt
package io.github.pshegger.infobook

import android.app.Application
import android.preference.PreferenceManager
import com.google.gson.GsonBuilder
import com.squareup.okhttp.Interceptor
import com.squareup.okhttp.OkHttpClient
import com.squareup.okhttp.Response
import retrofit.GsonConverterFactory
import retrofit.Retrofit
import java.util.*

/**
 * Created by pshegger on 2015. 11. 09..
 */
class InfoBookApplication : Application() {
    companion object {
        val RNG = Random()
        lateinit var instance: InfoBookApplication
    }

    var apiKey: String = ""
        get() = field
        set(value) {
            field = value
            PreferenceManager.getDefaultSharedPreferences(this)
                    .edit()
                    .putString(Constants.Preferences.API_KEY, value)
                    .apply()
            service = getNewApiService(value)
        }

    var service: GWApiService = getNewApiService()

    override fun onCreate() {
        super.onCreate()

        val prefs = PreferenceManager.getDefaultSharedPreferences(this)
        apiKey = prefs.getString(
                Constants.Preferences.API_KEY,
                "3674D0F3-B7A8-2242-B4C6-FE2E065C86A84C0DB899-DD85-490C-9986-1DA8E629CD21"
        )

        instance = this
    }

    private fun getNewApiService(newApiKey: String = apiKey): GWApiService {
        val client = OkHttpClient().apply {
            interceptors().add(object : Interceptor {
                override fun intercept(chain: Interceptor.Chain): Response {
                    val builder = chain.request().newBuilder()

                    builder.addHeader("Authorization", "Bearer $newApiKey")

                    return chain.proceed(builder.build())
                }
            })
        }

        val gson = GsonBuilder()
                .setDateFormat("yyyy-MM-dd'T'HH:mm:ssZ")
                .create()

        val retrofit = Retrofit.Builder()
                .client(client)
                .baseUrl(Constants.GW_API_URL)
                .addConverterFactory(GsonConverterFactory.create(gson))
                .build()

        return retrofit.create(GWApiService::class.java)
    }
}